File: Data Structures/TREE/Serialize and Deserialize Binary Tree.java
// https://leetcode.com/problems/serialize-and-deserialize-binary-tree/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if(root == null){
            return "X";
        }
        String l = serialize(root.left);
        String r = serialize(root.right);
        return root.val+","+l+","+r;
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        Queue<String> nodesLeft = new LinkedList<>();
        nodesLeft.addAll(Arrays.asList(data.split(",")));
        return deserializeHelper(nodesLeft);
    }
    
    public TreeNode deserializeHelper(Queue<String> nodesLeft){
        String val = nodesLeft.remove();
        if(val.equals("X")){
            return null;
        }
        TreeNode newNode = new TreeNode(Integer.valueOf(val));
        newNode.left = deserializeHelper(nodesLeft);
        newNode.right = deserializeHelper(nodesLeft);
        return newNode;
    }
    
}

File: Data Structures/TREE/Serialize and Deserialize Binary Tree.cpp
// https://practice.geeksforgeeks.org/problems/serialize-and-deserialize-a-binary-tree/1


void serialize(Node *root,vector<int> &A)
{
    if(root == NULL){
        A.push_back(-1);
        return;
    }
    
    A.push_back(root->data);
    serialize(root->left,A);
    serialize(root->right,A);
}

/*this function deserializes
 the serialized vector A*/
Node* solve(queue<int>& q){
    int x = q.front();
    q.pop();
    if(x==-1){
        return NULL;
    }
    
    Node* root = new Node(x);
    // cout<<root->data<<" \n";
    root->left = solve(q);
    root->right = solve(q);
    return root;
} 
 
Node * deSerialize(vector<int> &A)
{
   queue<int> q;
   for(auto x:A){
       q.push(x);
   }
   A.clear();
   Node* root = solve(q);
   return root;
}
